contracts/upgradeable_contracts/modules/interest/AAVEInterestERC20.sol

```diff
@@ -0,0 +1,240 @@
+    }
+
+    /**
+     * @dev Withdraws at least the given amount of tokens from the AAVE protocol.
+     * Only Omnibridge contract is allowed to call this method.
+     * Converts aTOKENs into _amount of TOKENs.
+     * @param _token address of the invested token contract.
+     * @param _amount minimal amount of tokens to withdraw.
+     */
+    function withdraw(address _token, uint256 _amount) external override onlyMediator {
+        InterestParams storage params = interestParams[_token];
+        uint256 invested = params.investedAmount;
+        uint256 redeemed = _safeWithdraw(_token, _amount > invested ? invested : _amount);
+        params.investedAmount = redeemed > invested ? 0 : invested - redeemed;
+        IERC20(_token).transfer(mediator, redeemed);
+    }
+
+    /**
+     * @dev Tells the current accumulated interest on the invested tokens, that can be withdrawn and payed to the interest receiver.
+     * @param _token address of the invested token contract.
+     * @return amount of accumulated interest.
+     */
+    function interestAmount(address _token) public view returns (uint256) {
+        InterestParams storage params = interestParams[_token];
+        (IAToken aToken, uint96 dust) = (params.aToken, params.dust);
+        uint256 balance = aToken.balanceOf(address(this));
+        // small portion of tokens are reserved for possible truncation/round errors
+        uint256 reserved = params.investedAmount.add(dust);
+        return balance > reserved ? balance - reserved : 0;
+    }
+
+    /**
+     * @dev Pays collected interest for the underlying token.
+     * Anyone can call this function.
+     * Earned interest is withdrawn and transferred to the specified interest receiver account.
+     * @param _token address of the invested token contract in which interest should be paid.
+     */
+    function payInterest(address _token) external {
+        InterestParams storage params = interestParams[_token];
+        uint256 interest = interestAmount(_token);
+        require(interest >= params.minInterestPaid);
+        _transferInterest(params.interestReceiver, address(_token), _safeWithdraw(_token, interest));
+    }
+
+    /**
+     * @dev Last-resort function for returning assets to the Omnibridge contract in case of some failures in the logic.
+     * Disables this contract and transfers locked tokens back to the mediator.
+     * Only owner is allowed to call this method.
+     * @param _token address of the invested token contract that should be disabled.
+     */
+    function forceDisable(address _token) external onlyOwner {
+        InterestParams storage params = interestParams[_token];
+        IAToken aToken = params.aToken;
+
+        uint256 aTokenBalance = aToken.balanceOf(address(this));
+        // try to redeem all aTokens
+        try lendingPool().withdraw(_token, aTokenBalance, mediator) {
+            uint256 balance = IERC20(_token).balanceOf(address(this));
```

Comment:
Does it make sense to pay the interest as part of the compounding disable process? If it is not possible, will it be beneficial to inform through an event how much it was earned? It will allow later to extract the interest from the withdrawn amount.

Reply:
I am convinced that such "emergency shutdown"-like function should be as simple as possible, that is why interest paying is a bit out of scope there. However, I see no problem with adding an extra event for easier recovery.

I have added a `ForceDisable` event to both interest implementations.